### DISCUSSION TITLE
UQTk install issue

### DISCUSSION THREAD

**Discussion by evasinha-pnnl** [sandialabs/UQTk#15] [Q&A] [answered]
I am running into problems when installing the current version of the UQTk. 

I tried installing it on NERSC perlmutter machine and PNNL compy and ran in to the same error at both machines. Here are the steps that I took:

```
module load cmake

git clone git@github.com:sandialabs/UQTk.git
mkdir build
cd ../build
../UQTk/config/config-gcc-Python.sh
make -j 8
```

`../UQTk/config/config-gcc-Python.sh`  step completes successfully but `make -j 8` fails with error: 
```
[ 94%] Linking CXX executable umbridge_model.x
CMakeFiles/umbridge_model.x.dir/umbridge_model.cpp.o: In function `void std::vector<std::thread, std::allocator<std::thread> >::_M_realloc_insert<httplib::ThreadPool::worker>(__gnu_cxx::__normal_iterator<std::thread*, std::vector<std::thread, std::allocator<std::thread> > >, httplib::ThreadPool::worker&&)':
umbridge_model.cpp:(.text._ZNSt6vectorISt6threadSaIS0_EE17_M_realloc_insertIJN7httplib10ThreadPool6workerEEEEvN9__gnu_cxx17__normal_iteratorIPS0_S2_EEDpOT_[_ZNSt6vectorISt6threadSaIS0_EE17_M_realloc_insertIJN7httplib10ThreadPool6workerEEEEvN9__gnu_cxx17__normal_iteratorIPS0_S2_EEDpOT_]+0xb2): undefined reference to `pthread_create'
CMakeFiles/umbridge_model.x.dir/umbridge_model.cpp.o: In function `std::_Function_handler<httplib::TaskQueue* (), httplib::Server::Server()::{lambda()#1}>::_M_invoke(std::_Any_data const&)':
umbridge_model.cpp:(.text._ZNSt17_Function_handlerIFPN7httplib9TaskQueueEvEZNS0_6ServerC4EvEUlvE_E9_M_invokeERKSt9_Any_data[_ZNSt17_Function_handlerIFPN7httplib9TaskQueueEvEZNS0_6ServerC4EvEUlvE_E9_M_invokeERKSt9_Any_data]+0xf1): undefined reference to `pthread_create'
collect2: error: ld returned 1 exit status
make[2]: *** [examples/tmcmc_umbridge/umbridge_model.x] Error 1
make[1]: *** [examples/tmcmc_umbridge/CMakeFiles/umbridge_model.x.dir/all] Error 2
make: *** [all] Error 2
```
Can you advice on how to get past this issue?

**Comment by bjdebus** [comment, marked as answer]
Hi @evasinha , 

Thanks for using UQTk and for reaching out about this issue. It looks like you are having a linking issue. Can you try to add the following line 
target_link_libraries (umbridge_model.x pthread)
after line 15 in the file examples/tmcmc_umbridge/CMakeLists.txt (with credit to @PieterjanRobbe for this suggestion).

After you do that, remove your current build directory, make a new one, and redo the config and make step.

If that does not fix it, you can disable the umbridge example by commenting out line 16 in examples/CMakeLists.txt (and starting with a new, clean build directory again).

Let us know if that works, so we can make it a permanent fix.
Thanks!
Bert

**Reply by evasinha-pnnl** [reply]
@bjdebus and @PieterjanRobbe Thank you very much for your help in resolving this issue. After adding the line described above, I was able to build, test, and install UQTk.

**Comment by PieterjanRobbe** [comment]
Hi @evasinha, I can replicate your issue on Perlmutter. The suggestion above seems to work for me, and tests are running fine now:

```
(nersc-python) 05:25 PM [pieterja@login23] ~/software/UQTk/build> ctest
...
100% tests passed, 0 tests failed out of 31

Total Test time (real) =  16.70 sec
```

**Reply by evasinha-pnnl** [reply]
Yes, it worked for me also on Perlmutter and Compy both.

**Comment by bjdebus** [comment]
Thanks @evasinha for testing this bugfix and for confirming that it works. We will fix this in the main branch too and push it to github soon. Regards, Bert
